File: typescript/ts_sdk/src/authentication/did-wba-authenticator.ts
import { readFile } from 'node:fs/promises';

import { generateAuthHeader } from './did-wba.js';
import { generateHttpSignatureHeaders } from './http-signatures.js';
import { AuthMode, type DidDocument } from './types.js';

export class DIDWbaAuthHeader {
  private didDocumentCache?: DidDocument;

  private readonly tokens = new Map<string, string>();

  constructor(
    private readonly didDocumentPath: string,
    private readonly privateKeyPath: string,
    private readonly authMode: AuthMode = AuthMode.HttpSignatures
  ) {}

  async getAuthHeaders(
    serverUrl: string,
    forceNew = false,
    method = 'GET',
    headers?: Record<string, string>,
    body?: Uint8Array | string
  ): Promise<Record<string, string>> {
    const domain = extractDomain(serverUrl);
    if (!forceNew) {
      const token = this.tokens.get(domain);
      if (token) {
        return { Authorization: `Bearer ${token}` };
      }
    }

    const [didDocument, privateKeyPem] = await Promise.all([
      this.loadDidDocument(),
      readFile(this.privateKeyPath, 'utf8'),
    ]);

    if (this.authMode === AuthMode.LegacyDidWba) {
      return {
        Authorization: generateAuthHeader(didDocument, domain, privateKeyPem, '1.1'),
      };
    }

    return generateHttpSignatureHeaders(
      didDocument,
      serverUrl,
      method,
      privateKeyPem,
      headers,
      body
    );
  }

  async getAuthHeader(
    serverUrl: string,
    forceNew = false,
    method = 'GET',
    headers?: Record<string, string>,
    body?: Uint8Array | string
  ): Promise<Record<string, string>> {
    return this.getAuthHeaders(serverUrl, forceNew, method, headers, body);
  }

  updateToken(serverUrl: string, headers: Record<string, string>): string | undefined {
    const domain = extractDomain(serverUrl);
    const authenticationInfo = getHeaderCaseInsensitive(headers, 'Authentication-Info');
    if (authenticationInfo) {
      const parsed = parseAuthenticationInfo(authenticationInfo);
      const accessToken = parsed.access_token;
      const tokenType = parsed.token_type ?? 'Bearer';
      if (accessToken && tokenType.toLowerCase() === 'bearer') {
        this.tokens.set(domain, accessToken);
        return accessToken;
      }
    }
    const authorization = getHeaderCaseInsensitive(headers, 'Authorization');
    if (authorization?.startsWith('Bearer ')) {
      const token = authorization.slice(7);
      this.tokens.set(domain, token);
      return token;
    }
    return undefined;
  }

  clearToken(serverUrl: string): void {
    this.tokens.delete(extractDomain(serverUrl));
  }

  clearAllTokens(): void {
    this.tokens.clear();
  }

  shouldRetryAfter401(responseHeaders: Record<string, string>): boolean {
    const wwwAuthenticate = getHeaderCaseInsensitive(responseHeaders, 'WWW-Authenticate');
    if (!wwwAuthenticate) {
      return false;
    }
    const challenge = parseWwwAuthenticate(wwwAuthenticate);
    if (challenge.nonce) {
      return true;
    }
    return !['invalid_did', 'invalid_verification_method', 'forbidden_did'].includes(
      challenge.error ?? ''
    );
  }

  async getChallengeAuthHeaders(
    serverUrl: string,
    responseHeaders: Record<string, string>,
    method = 'GET',
    headers?: Record<string, string>,
    body?: Uint8Array | string
  ): Promise<Record<string, string>> {
    const wwwAuthenticate = getHeaderCaseInsensitive(responseHeaders, 'WWW-Authenticate');
    const acceptSignature = getHeaderCaseInsensitive(responseHeaders, 'Accept-Signature');
    const challenge = wwwAuthenticate ? parseWwwAuthenticate(wwwAuthenticate) : {};
    const coveredComponents = normalizeCoveredComponents(
      acceptSignature ? parseAcceptSignature(acceptSignature) : undefined,
      headers,
      body
    );

    const [didDocument, privateKeyPem] = await Promise.all([
      this.loadDidDocument(),
      readFile(this.privateKeyPath, 'utf8'),
    ]);

    if (this.authMode === AuthMode.LegacyDidWba) {
      return {
        Authorization: generateAuthHeader(
          didDocument,
          extractDomain(serverUrl),
          privateKeyPem,
          '1.1',
          {
            nonce: challenge.nonce,
          }
        ),
      };
    }

    return generateHttpSignatureHeaders(
      didDocument,
      serverUrl,
      method,
      privateKeyPem,
      headers,
      body,
      {
        nonce: challenge.nonce,
        coveredComponents,
      }
    );
  }

  async getChallengeAuthHeader(
    serverUrl: string,
    responseHeaders: Record<string, string>,
    method = 'GET',
    headers?: Record<string, string>,
    body?: Uint8Array | string
  ): Promise<Record<string, string>> {
    return this.getChallengeAuthHeaders(serverUrl, responseHeaders, method, headers, body);
  }

  private async loadDidDocument(): Promise<DidDocument> {
    if (!this.didDocumentCache) {
      this.didDocumentCache = JSON.parse(
        await readFile(this.didDocumentPath, 'utf8')
      ) as DidDocument;
    }
    return this.didDocumentCache;
  }
}

function extractDomain(serverUrl: string): string {
  try {
    return new URL(serverUrl).hostname;
  } catch {
    return serverUrl;
  }
}

function getHeaderCaseInsensitive(
  headers: Record<string, string>,
  name: string
): string | undefined {
  const target = name.toLowerCase();
  return Object.entries(headers).find(([key]) => key.toLowerCase() === target)?.[1];
}

function parseAuthenticationInfo(value: string): Record<string, string> {
  return value
    .split(',')
    .map((item) => item.trim())
    .filter(Boolean)
    .reduce<Record<string, string>>((result, item) => {
      const [key, rawValue] = item.split('=', 2);
      if (key && rawValue) {
        result[key.trim()] = rawValue.trim().replace(/^"|"$/g, '');
      }
      return result;
    }, {});
}

function parseWwwAuthenticate(value: string): Record<string, string> {
  const normalized = value.replace(/^DIDWba\s+/i, '').trim();
  const matches = [...normalized.matchAll(/([\w-]+)=("[^"]*"|[^,]+)/g)];
  return matches.reduce<Record<string, string>>((result, match) => {
    result[match[1]] = match[2].trim().replace(/^"|"$/g, '');
    return result;
  }, {});
}

function parseAcceptSignature(value: string): string[] {
  return [...value.matchAll(/"([^"]+)"/g)].map((match) => match[1]);
}

function normalizeCoveredComponents(
  coveredComponents: string[] | undefined,
  headers: Record<string, string> | undefined,
  body: Uint8Array | string | undefined
): string[] | undefined {
  if (!coveredComponents) {
    return undefined;
  }
  const normalizedHeaders = Object.fromEntries(
    Object.entries(headers ?? {}).map(([key, value]) => [key.toLowerCase(), value])
  );
  const bodyPresent =
    body !== undefined &&
    !(
      (typeof body === 'string' && body.length === 0) ||
      (body instanceof Uint8Array && body.byteLength === 0)
    );

  return coveredComponents.filter((component) => {
    const normalized = component.toLowerCase();
    if (normalized === 'content-digest' && !bodyPresent) {
      return false;
    }
    if (
      normalized === 'content-length' &&
      !bodyPresent &&
      !('content-length' in normalizedHeaders)
    ) {
      return false;
    }
    if (normalized === 'content-type' && !('content-type' in normalizedHeaders)) {
      return false;
    }
    if (
      !normalized.startsWith('@') &&
      normalized !== 'content-length' &&
      normalized !== 'content-digest' &&
      !(normalized in normalizedHeaders)
    ) {
      return false;
    }
    return true;
  });
}

File: anp/authentication/did_wba_authenticator.py
# AgentConnect: https://github.com/agent-network-protocol/AgentConnect
# Author: GaoWei Chang
# Email: chgaowei@gmail.com
# Website: https://agent-network-protocol.com/
#
# This project is open-sourced under the MIT License. For details, please see the LICENSE file.

import logging
import re
from typing import Any, Dict, List, Optional
from urllib.parse import urlparse

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec, ed25519

from .did_wba import generate_auth_header
from .http_signatures import generate_http_signature_headers


class DIDWbaAuthHeader:
    """Client helper for DID-WBA authentication headers and token caching."""

    def __init__(
        self,
        did_document_path: str,
        private_key_path: str,
        auth_mode: str = "http_signatures",
    ):
        """Initialize the DID authentication client.

        Args:
            did_document_path: Path to the DID document file.
            private_key_path: Path to the private key file.
            auth_mode: Authentication mode. Supported values are:
                - "http_signatures": Use HTTP Message Signatures (default)
                - "legacy_didwba": Use the legacy Authorization header scheme
                - "auto": Alias of "http_signatures" for forward compatibility
        """
        self.did_document_path = did_document_path
        self.private_key_path = private_key_path
        self.auth_mode = auth_mode

        self.did_document = None
        self.tokens: Dict[str, str] = {}

        logging.info("DIDWbaAuthHeader initialized with auth mode %s", auth_mode)

    def _get_domain(self, server_url: str) -> str:
        """Extract domain from URL."""
        parsed_url = urlparse(server_url)
        return parsed_url.netloc.split(":")[0]

    @staticmethod
    def _get_header_case_insensitive(
        headers: Dict[str, str], name: str
    ) -> Optional[str]:
        lower_name = name.lower()
        for key, value in headers.items():
            if key.lower() == lower_name:
                return value
        return None

    def _load_did_document(self) -> Dict[str, Any]:
        """Load and cache the DID document from disk."""
        if self.did_document is not None:
            return self.did_document

        try:
            with open(self.did_document_path, "r", encoding="utf-8") as file_obj:
                import json

                self.did_document = json.load(file_obj)
            logging.info("Loaded DID document from %s", self.did_document_path)
            return self.did_document
        except Exception as exc:
            logging.error("Error loading DID document: %s", exc)
            raise

    def _load_private_key(self) -> Any:
        """Load the private key from disk."""
        try:
            with open(self.private_key_path, "rb") as file_obj:
                private_key_data = file_obj.read()
            return serialization.load_pem_private_key(private_key_data, password=None)
        except Exception as exc:
            logging.error("Error loading private key: %s", exc)
            raise

    def _sign_callback(self, content: bytes, method_fragment: str) -> bytes:
        """Sign arbitrary bytes with the loaded private key."""
        try:
            private_key = self._load_private_key()
            if isinstance(private_key, ec.EllipticCurvePrivateKey):
                signature = private_key.sign(content, ec.ECDSA(hashes.SHA256()))
            elif isinstance(private_key, ed25519.Ed25519PrivateKey):
                signature = private_key.sign(content)
            else:
                raise TypeError(
                    f"Unsupported private key type: {type(private_key).__name__}"
                )
            logging.debug(
                "Signed authentication content with method fragment %s",
                method_fragment,
            )
            return signature
        except Exception as exc:
            logging.error("Error signing content: %s", exc)
            raise

    def _generate_legacy_auth_header(self, domain: str) -> Dict[str, str]:
        """Generate a legacy DIDWba Authorization header."""
        did_document = self._load_did_document()
        auth_header = generate_auth_header(did_document, domain, self._sign_callback)
        logging.info("Generated legacy DIDWba Authorization header for %s", domain)
        return {"Authorization": auth_header}

    def _generate_http_signature_header(
        self,
        server_url: str,
        method: str,
        headers: Optional[Dict[str, str]] = None,
        body: Any = None,
    ) -> Dict[str, str]:
        """Generate HTTP Message Signatures headers for a request."""
        did_document = self._load_did_document()
        auth_headers = generate_http_signature_headers(
            did_document=did_document,
            request_url=server_url,
            request_method=method,
            sign_callback=self._sign_callback,
            headers=headers,
            body=body,
        )
        logging.info("Generated HTTP signature headers for %s %s", method, server_url)
        return auth_headers

    def get_auth_header(
        self,
        server_url: str,
        force_new: bool = False,
        method: str = "GET",
        headers: Optional[Dict[str, str]] = None,
        body: Any = None,
    ) -> Dict[str, str]:
        """Get authentication headers for a request.

        When a token is cached and force_new is False, this method returns
        an Authorization Bearer header. Otherwise it generates either new HTTP
        signature headers or a legacy DIDWba Authorization header depending on
        auth_mode.

        Args:
            server_url: Exact request URL, including the final path and query string.
            force_new: If True, bypass the cached Bearer token and sign a new request.
            method: Exact HTTP method for the outgoing request.
            headers: Request headers that should be covered when applicable.
            body: Exact request bytes or string that will be sent on the wire.
        """
        domain = self._get_domain(server_url)
        if domain in self.tokens and not force_new:
            token = self.tokens[domain]
            logging.info("Using cached Bearer token for %s", domain)
            return {"Authorization": f"Bearer {token}"}

        normalized_mode = self.auth_mode.lower()
        if normalized_mode in {"http_signatures", "auto"}:
            return self._generate_http_signature_header(
                server_url=server_url,
                method=method,
                headers=headers,
                body=body,
            )
        if normalized_mode == "legacy_didwba":
            return self._generate_legacy_auth_header(domain)
        raise ValueError(
            "auth_mode must be one of: http_signatures, legacy_didwba, auto"
        )

    @staticmethod
    def _parse_authentication_info(header_value: str) -> Dict[str, str]:
        """Parse Authentication-Info key/value pairs."""
        parts = re.findall(r'(\w+)=("[^"]*"|[^,]+)', header_value)
        result: Dict[str, str] = {}
        for key, raw_value in parts:
            value = raw_value.strip()
            if value.startswith('"') and value.endswith('"'):
                value = value[1:-1]
            result[key] = value
        return result

    @staticmethod
    def _parse_www_authenticate(header_value: str) -> Dict[str, str]:
        """Parse DIDWba WWW-Authenticate challenge fields."""
        value = header_value.strip()
        if " " in value:
            scheme, remainder = value.split(" ", 1)
            if scheme.lower() == "didwba":
                value = remainder
        parts = re.findall(r'([\w-]+)=("[^"]*"|[^,]+)', value)
        result: Dict[str, str] = {}
        for key, raw_value in parts:
            parsed_value = raw_value.strip()
            if parsed_value.startswith('"') and parsed_value.endswith('"'):
                parsed_value = parsed_value[1:-1]
            result[key] = parsed_value
        return result

    @staticmethod
    def _parse_accept_signature(header_value: str) -> List[str]:
        """Parse covered components from an Accept-Signature header."""
        return re.findall(r'"([^"]+)"', header_value)

    @staticmethod
    def _normalize_covered_components(
        covered_components: Optional[List[str]],
        headers: Optional[Dict[str, str]],
        body: Any,
    ) -> Optional[List[str]]:
        """Drop challenge components that cannot be satisfied for this request."""
        if covered_components is None:
            return None

        normalized_headers = {key.lower(): value for key, value in (headers or {}).items()}
        body_present = body not in (None, b"", "")
        normalized_components: List[str] = []
        for component in covered_components:
            component_lower = component.lower()
            if component_lower == "content-digest" and not body_present:
                continue
            if (
                component_lower == "content-length"
                and not body_present
                and "content-length" not in normalized_headers
            ):
                continue
            if component_lower == "content-type" and "content-type" not in normalized_headers:
                continue
            if (
                not component_lower.startswith("@")
                and component_lower not in normalized_headers
                and component_lower != "content-length"
                and component_lower != "content-digest"
            ):
                continue
            normalized_components.append(component)
        return normalized_components

    def should_retry_after_401(self, response_headers: Dict[str, str]) -> bool:
        """Return whether a 401 response should trigger one authentication retry."""
        www_authenticate = self._get_header_case_insensitive(
            response_headers,
            "WWW-Authenticate",
        )
        if not www_authenticate:
            return False

        challenge = self._parse_www_authenticate(www_authenticate)
        error = challenge.get("error", "").lower()
        if challenge.get("nonce"):
            return True
        return error not in {
            "invalid_did",
            "invalid_verification_method",
            "forbidden_did",
        }

    def get_challenge_auth_header(
        self,
        server_url: str,
        response_headers: Dict[str, str],
        method: str = "GET",
        headers: Optional[Dict[str, str]] = None,
        body: Any = None,
    ) -> Dict[str, str]:
        """Generate DID authentication headers for a 401 challenge response."""
        www_authenticate = self._get_header_case_insensitive(
            response_headers,
            "WWW-Authenticate",
        )
        accept_signature = self._get_header_case_insensitive(
            response_headers,
            "Accept-Signature",
        )
        challenge = (
            self._parse_www_authenticate(www_authenticate)
            if www_authenticate
            else {}
        )
        covered_components = self._normalize_covered_components(
            self._parse_accept_signature(accept_signature)
            if accept_signature
            else None,
            headers,
            body,
        )
        nonce = challenge.get("nonce")

        normalized_mode = self.auth_mode.lower()
        if normalized_mode in {"http_signatures", "auto"}:
            did_document = self._load_did_document()
            auth_headers = generate_http_signature_headers(
                did_document=did_document,
                request_url=server_url,
                request_method=method,
                sign_callback=self._sign_callback,
                headers=headers,
                body=body,
                nonce=nonce,
                covered_components=covered_components,
            )
            logging.info(
                "Generated HTTP challenge response headers for %s %s",
                method,
                server_url,
            )
            return auth_headers

        if normalized_mode == "legacy_didwba":
            did_document = self._load_did_document()
            auth_header = generate_auth_header(
                did_document,
                self._get_domain(server_url),
                self._sign_callback,
                nonce=nonce,
            )
            logging.info(
                "Generated legacy challenge response header for %s",
                server_url,
            )
            return {"Authorization": auth_header}

        raise ValueError(
            "auth_mode must be one of: http_signatures, legacy_didwba, auto"
        )

    def update_token(self, server_url: str, headers: Dict[str, str]) -> Optional[str]:
        """Update the cached token from response headers."""
        domain = self._get_domain(server_url)

        authentication_info = self._get_header_case_insensitive(
            headers, "Authentication-Info"
        )
        if authentication_info:
            info = self._parse_authentication_info(authentication_info)
            access_token = info.get("access_token")
            token_type = info.get("token_type", "Bearer")
            if access_token and token_type.lower() == "bearer":
                self.tokens[domain] = access_token
                logging.info(
                    "Updated token from Authentication-Info for %s", domain
                )
                return access_token

        auth_header = self._get_header_case_insensitive(headers, "Authorization")
        if auth_header and auth_header.lower().startswith("bearer "):
            token = auth_header[7:]
            self.tokens[domain] = token
            logging.info("Updated token from Authorization header for %s", domain)
            return token

        logging.debug("No bearer token found in response headers for %s", domain)
        return None

    def clear_token(self, server_url: str) -> None:
        """Clear token for the specified domain."""
        domain = self._get_domain(server_url)
        if domain in self.tokens:
            del self.tokens[domain]
            logging.info("Cleared token for %s", domain)

    def clear_all_tokens(self) -> None:
        """Clear all cached tokens."""
        self.tokens.clear()
        logging.info("Cleared all cached tokens")
